sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/TransactionalBatch.java

```diff
@@ -0,0 +1,351 @@
+    }
+
+    /**
+     * Adds an operation to replace an item into the batch.
+     *
+     * @param <T> The type of item to be replaced.
+     *
+     * @param id The unique id of the item.
+     * @param item A JSON serializable object that must contain an id property.
+     * @param requestOptions The options for the item request.
+     *
+     * @return The transactional batch instance with the operation added.
+     */
+    public <T> CosmosItemOperation replaceItemOperation(
+        String id, T item, TransactionalBatchItemRequestOptions requestOptions) {
+
+        checkNotNull(id, "expected non-null id");
+        checkNotNull(item, "expected non-null item");
+        if (requestOptions == null) {
+            requestOptions = new TransactionalBatchItemRequestOptions();
+        }
+
+        ItemBatchOperation<T> operation = new ItemBatchOperation<T>(
+            CosmosItemOperationType.REPLACE,
+            id,
+            this.getPartitionKeyValue(),
+            requestOptions.toRequestOptions(),
+            item
+        );
+
+        this.operations.add(operation);
+
+        return operation;
+    }
+
+    /**
+     * Adds an operation to upsert an item into the batch.
+     *
+     * @param item A JSON serializable object that must contain an id property.
+     * @param <T> The type of item to be upserted.
+     *
+     * @return The transactional batch instance with the operation added.
+     */
+    public <T> CosmosItemOperation upsertItemOperation(T item) {
+        checkNotNull(item, "expected non-null item");
+        return this.upsertItemOperation(item, new TransactionalBatchItemRequestOptions());
+    }
+
+    /**
+     * Adds an operation to upsert an item into the batch.
+     *
+     * @param <T> The type of item to be upserted.
+     *
+     * @param item A JSON serializable object that must contain an id property.
+     * @param requestOptions The options for the item request.
+     *
+     * @return The transactional batch instance with the operation added.
+     */
+    public <T> CosmosItemOperation upsertItemOperation(T item, TransactionalBatchItemRequestOptions requestOptions) {
+
+        checkNotNull(item, "expected non-null item");
+        if (requestOptions == null) {
+            requestOptions = new TransactionalBatchItemRequestOptions();
+        }
+
+        ItemBatchOperation<T> operation = new ItemBatchOperation<T>(
+            CosmosItemOperationType.UPSERT,
+            null,
+            this.getPartitionKeyValue(),
+            requestOptions.toRequestOptions(),
+            item
+        );
+
+        this.operations.add(operation);
+
+        return operation;
+    }
+
+    /**
+     * Return the list of operation in an unmodifiable instance  so no one can change it in the down path.
+     *
+     * @return The list of operations which are to be executed.
+     */
+    List<ItemBatchOperation<?>> getOperationsInternal() {
+        return UnmodifiableList.unmodifiableList(operations);
```

Comment:
duplicate.
you can rely on getOperations() public method.

Reply:
Yeah changed it to use CosmosItemOperation everywhere except when serializing it.